src/monodroid/jni/monodroid-glue.c

```diff
@@ -704,55 +723,104 @@ typedef void* (*mono_mkbundle_init_ptr) (void (*)(const MonoBundledAssembly **),
 mono_mkbundle_init_ptr mono_mkbundle_init;
 void (*mono_mkbundle_initialize_mono_api) (const BundleMonoAPI *info);
 
-static void
-setup_bundled_app (const char *libappso)
+static char*
+get_full_dso_path (const char *base_dir, const char *dso_path, mono_bool *needs_free)
 {
-	void *libapp;
+	if (needs_free != NULL)
+		*needs_free = FALSE;
+	if (!dso_path)
+		return NULL;
 
-	libapp = dlopen (libappso, RTLD_LAZY);
+	if (base_dir == NULL || dso_path [0] == MONODROID_PATH_SEPARATOR_CHAR)
+		return (char*)dso_path; // Absolute path, can't do much with it
```

Comment:
*Especially* considering the eventual C++ migration, is this really a good idea, casting away `const` like this?

Should we instead make the `dso_path` non-`const` for sanity?

Reply:
Yes and no. `dso_path` should be a `const` - we shouldn't modify it in any way. OTOH, the return value must be non-const since we **may** allocate it. In this case a conscious cast like this is OK - in the presence of the `needs_free` flag we make the intent explicit. The parameter `const` status is a promise we won't modify the parameter.

Reply:
I think `dso_path` shouldn't be *const* here, as by the cast before returning it you remove the *const* and have no control of what happens with the returned value afterwards. And because the return value can be allocated and thus we need to call `free` on it later, you cannot return *const* char*. (or you can, but have to cast it when calling free, which I find less harmful as it is at the end of the life of it)

IIRC, once we switch to C++ you might return *const* pointer, as the delete operator can be used with the const pointer.

Reply:
I think it's safer (now) to return `char*` and do the cast inside the function than rely on the caller to perform the cast - the caller has the boolean flag to tell them when to free the returned value. Granted, it's not ideal, but within the realm of C there's no good way to handle this situation. In C++ we can do what @radekdoulik suggests or use a wrapper class which will know what to do with the stored value.